### PR TITLE
Improve Bloom Filter aggregations in Sparse Joins

### PR DESCRIPTION
**Pull request by anish749**
BloomFilters used in Sparse Join implementations use Algebird aggregators. This step incurs a good cost in terms of memory and cpu when using `.aggregate(com.twitter.algebird.Aggregator)` because of the way BloomFilters are modelled in Algebird. The aggregate goes as prepare -> sum -> present. For BloomFilterAggregator, the prepare step creates a BFMonoid for each element, and then the sum adds up the monoids. This means that all the keys becomes its own monoid, which takes up a lot of intermediate space. (Around 8 times more space than the keys itself). This can be seen when we expand the transforms in Dataflow.

screenshot 2019-01-31 at 19 48 32
Prepare step of aggregation

screenshot 2019-01-31 at 19 50 24
Sum operation of aggregation

The vCPU time for the sum operation is around 14 hrs, which is quite high.

Changing this to use our own combine functions, provides significant performance improvements.
`.aggregate(bfAggregator.monoid.zero)(_ + _, _ ++ _)`
screenshot 2019-01-31 at 19 54 09

By removing prepare, the intermediate step which was having an output more than 8 times the size of the keys is now removed. The vCPU time has also come down to 5 hrs. (35% of what it was previously)

We noticed up 20% less overall CPU usage on one of our jobs with this change.

Also in this PR.

Added the sparse lookups in a transform to make it look pretty in DataFlow UI.

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 1]
```diff
@@ -488,14 +487,14 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
...
     val numKeysPerPartition = if (bfSettings.numBFs == 1) thisNumKeys.toInt else bfSettings.capacity
     val n = bfSettings.numBFs
+    val width = BloomFilter.optimalWidth(numKeysPerPartition, fpProb).get
+    val numHashes = BloomFilter.optimalNumHashes(numKeysPerPartition, width)
+    val bfAggregator = BloomFilterAggregator[K](numHashes, width)
     self
-      .partition(n, _._1.hashCode() % n)
+      .partition(n, { case (key, _) => key.hashCode() % n })
       .map { me =>
-        val width = BloomFilter.optimalWidth(numKeysPerPartition, fpProb).get
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: 👍 moving this out

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 2]
```diff
@@ -488,14 +487,14 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
...
-        val numHashes = BloomFilter.optimalNumHashes(numKeysPerPartition, width)
-        val bfAggregator = BloomFilterAggregator[K](numHashes, width)
         me.keys
-          .aggregate(bfAggregator)
+          .aggregate(bfAggregator.monoid.zero)(_ + _, _ ++ _)
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: this is a really nice optimization! 👍

**Comment by anish749** [comment]
There are some more bugs.. I'll create another PR soon..!

**Comment by regadas** [comment]
@anish749 interesting ... around this PR or on other stuff?

**Pull request merged**